From ecc11bb61e6d2b7075efe977a77fe93a583d451c Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:44 -0400
Subject: [fs] btrfs: check if items are ordered when a leaf is marked dirty

Message-id: <1407179109-13714-338-git-send-email-sandeen@redhat.com>
Patchwork-id: 87233
O-Subject: [RHEL7.1 PATCH 337/422] Btrfs: check if items are ordered when a leaf is marked dirty
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Filipe Manana <fdmanana@gmail.com>

commit 1f21ef0a34c37de0daf66fce7141b23b88630f29
Author: Filipe Manana <fdmanana@gmail.com>
Date:   Wed Apr 9 15:37:06 2014 +0100

    Btrfs: check if items are ordered when a leaf is marked dirty

    To ease finding bugs during development related to modifying btree leaves
    in such a way that it makes its items not sorted by key anymore. Since this
    is an expensive check, it's only enabled if CONFIG_BTRFS_FS_CHECK_INTEGRITY
    is set, which isn't meant to be enabled for regular users.

    Signed-off-by: Filipe David Borba Manana <fdmanana@gmail.com>
    Reviewed-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index c2544fd..237d704 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -3740,6 +3740,12 @@ void btrfs_mark_buffer_dirty(struct extent_buffer *buf)
   __percpu_counter_add(&root->fs_info->dirty_metadata_bytes,
          buf->len,
          root->fs_info->dirty_metadata_batch);
+#ifdef CONFIG_BTRFS_FS_CHECK_INTEGRITY
+ if (btrfs_header_level(buf) == 0 && check_leaf(root, buf)) {
+  btrfs_print_leaf(root, buf);
+  ASSERT(0);
+ }
+#endif
 }
 
 static void __btrfs_btree_balance_dirty(struct btrfs_root *root,
-- 
1.7.1